drivers/hmis/app/models/hmis/entity_access_loader_factory.rb

```diff
@@ -0,0 +1,88 @@
+####
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+##
+
+# Factory to resolve the user access loader for a given entity. It also resolves
+# the entity that should be passed to the loader. A resolver blocker is used to
+# traverse associations as, in the case of graphql, another data loader would be
+# used to resolve the association.
+#
+# It's up to the caller to actually invoke the loader with the resolved entity to
+# perform the access check.
+#
+# @example
+#   factory = Hmis::EntityAccessLoaderFactory.new(user)
+#   loader, entity = factory.perform(project) do |record, association|
+#     # resolve an entity association. In graphql, a data loader would be used
+#     association.send(record)
+#   end
+#   # resolve a user's permission on project through the loader. Loader.fetch() would
+#   # be used to called for many projects
+#   allowed = loader.fetch_one(entity, :can_do_something)
+class Hmis::EntityAccessLoaderFactory
+  # @param user [Hmis::User]
+  def initialize(user)
+    @user = user
+  end
+
+  # Given an entity, return a tuple of the loader and the entity to check against
+  #
+  # @param entity [#entity_record] active record entity
+  # @yieldparam [#entity_record] record
+  # @yieldparam [Symbol] association name
+  # @yieldreturn [#resolved, nil] the association (record.association)
+  # @return [Array<Hmis::BaseLoader, #entity>]
+  def perform(entity, safety: 0, &block)
+    check_safety(safety)
+
+    loader = case entity
+    when Hmis::Hud::Client
+      # The client access loader can handle new clients directly
+      Hmis::Hud::ClientAccessLoader
+    when Hmis::Hud::Project
+      Hmis::Hud::ProjectAccessLoader if entity.persisted?
+    when Hmis::Hud::Organization
+      Hmis::Hud::OrganizationAccessLoader if entity.persisted?
+    when GrdaWarehouse::DataSource
+      Hmis::DataSourceAccessLoader if entity.persisted?
+    end
+    if loader
+      [loader.new(@user), entity]
+    else
+      # recurse
+      resolve_association(entity, safety: safety + 1, &block)
+    end
+  end
+
+  # follow the association chain for entities that don't have their own data loader
+  # or for entities that are not yet persisted
+  def resolve_association(entity, safety:, &block)
+    check_safety(safety)
+
+    resolved = case entity
+    when Hmis::File
+      # regardless of if it's a new record, use the file's client
+      # FIXME: this used to use the optional file.enrollment, confirm this doesn't matter
+      block.call(entity, :client)
+    when Hmis::Hud::HmisService, Hmis::Hud::Service
+      # This will cascade to enrollment.project. We go through enrollment since
+      # service.project is just a method on these models
+      block.call(entity, :enrollment)
```

Comment:
Let's discuss this one. Nearly all of the HUD models have a project through enrollment, and the reason the project access is defined as a method is because of the WIP issue. With the views, it can probably be removed.